### PR TITLE
Create new connector for SAP HANA

### LINKED ISSUE #44: [FEATURE] Add support for SAP HANA
**What would you like to be added**:
Support for ingesting metadata from SAP HANA into Google Data Catalog.

**Why is this needed**:
SAP HANA is a high-performance in-memory database that speeds data-driven, real-time decisions and actions. SAP HANA customers might benefit from Data Catalog's powerful metadata management capabilities in order to discover and better manage their metadata.

### PR DESCRIPTION
**Pull request by mesmacosta**
**- What I did**
Created the SAP HANA connector, that ingests tables and schemas with their metadata into Google Data Catalog.

**- How I did it**
Created the initial code, and tested the ingestion on a SAP HANA Express edition.

**- How to verify it**
Run the connector in a SAP HANA Database.

**- Description for the changelog**
New connector for SAP HANA. Fixes #44

### CODE REVIEW

File: google-datacatalog-saphana-connector/setup.py [hunk 1]
```diff
@@ -0,0 +1,49 @@
...
+            'google-datacatalog-saphana-connector = google.datacatalog_connectors.saphana:main',
+        ],
+    },
+    include_package_data=True,
+    install_requires=('hdbcli', 'google-datacatalog-rdbms-connector>=0.9.0'),
+    setup_requires=('pytest-runner',),
+    tests_require=('pytest-cov', 'google-datacatalog-connectors-commons-test>=0.6.0'),
+    classifiers=(
+        release_status,
+        'Programming Language :: Python :: 3.7',
```
Comment (ricardolsmendes) [inline google-datacatalog-saphana-connector/setup.py:45]:
```suggestion
        'Programming Language :: Python :: 3.6',
```
Is it possible to use 3.6 to keep it consistent with the README file?

File: google-datacatalog-saphana-connector/src/google/datacatalog_connectors/saphana/datacatalog_cli.py [hunk 2]
```diff
@@ -0,0 +1,98 @@
...
+        return os.path.join(os.path.dirname(os.path.abspath(__file__)),
+                            'config/metadata_definition.json')
+
+    def _get_query_path(self, args):
+        return os.path.join(os.path.dirname(os.path.abspath(__file__)),
+                            'config/metadata_query.sql')
+
+    def _parse_args(self, argv):
+        parser = argparse.ArgumentParser(
+            description='Command line to sync saphana '
```
Comment (ricardolsmendes) [inline google-datacatalog-saphana-connector/src/google/datacatalog_connectors/saphana/datacatalog_cli.py:55]:
```suggestion
            description='Command line to sync SAP Hana '
```
The official product name is `SAP Hana`, so I recommend using it in any kind of documentation and messages. WDYT? 🤔

File: google-datacatalog-saphana-connector/src/google/datacatalog_connectors/saphana/datacatalog_cli.py [hunk 3]
```diff
@@ -0,0 +1,98 @@
...
+                            'Cloud Datacatalog')
+        parser.add_argument(
+            '--saphana-host',
+            help='Your saphana server host, this is required even'
+            ' for the raw_metadata_csv,'
+            ' so we are able to map the created entries'
+            ' resource with the sap hana host',
+            required=True)
+        parser.add_argument('--saphana-user',
+                            help='Your saphana credentials user')
```
Comment (ricardolsmendes) [inline google-datacatalog-saphana-connector/src/google/datacatalog_connectors/saphana/datacatalog_cli.py:76]:
```suggestion
                            help='Your SAP Hana credentials user')
```

File: google-datacatalog-saphana-connector/src/google/datacatalog_connectors/saphana/datacatalog_cli.py [hunk 4]
```diff
@@ -0,0 +1,98 @@
...
+        parser.add_argument('--saphana-pass',
+                            help='Your saphana credentials password')
+        parser.add_argument('--saphana-database',
+                            help='Your saphana database name')
+        parser.add_argument(
+            '--raw-metadata-csv',
+            help='Your raw metadata as a csv file, '
+            'can be either a local os GCS '
+            'path (If supplied ignores the sap hana server credentials)')
```
Comment (ricardolsmendes) [inline google-datacatalog-saphana-connector/src/google/datacatalog_connectors/saphana/datacatalog_cli.py:85]:
```suggestion
            'path (If supplied ignores the SAP Hana server credentials)')
```

File: google-datacatalog-saphana-connector/tests/google/datacatalog_connectors/saphana/test_data/metadata.json [hunk 5]
```diff
@@ -0,0 +1 @@
+{"schemas": [{"name": "saphana", "tables": [{"columns": [{"default_value": NaN, "name": "a", "nullable": "YES", "type": "integer"}], "name": "test_temp", "type": "BASE TABLE"}, {"columns": [{"default_value": NaN, "name": "uid", "nullable": "YES", "type": "integer"}, {"default_value": NaN, "name": "mid", "nullable": "NO", "type": "integer"}, {"default_value": NaN, "name": "message", "nullable": "YES", "type": "character varying"}, {"default_value": NaN, "name": "ptime", "nullable": "YES", "type": "date"}], "name": "messages", "type": "BASE TABLE"}, {"columns": [{"default_value": NaN, "name": "uid", "nullable": "NO", "type": "integer"}, {"default_value": NaN, "name": "name", "nullable": "YES", "type": "character varying"}], "name": "users", "type": "BASE TABLE"}]}]}
```
Comment (ricardolsmendes) [inline google-datacatalog-saphana-connector/tests/google/datacatalog_connectors/saphana/test_data/metadata.json:1]: `NaN` is not a valid JSON token AFAIK. Is it possible to replace it and strictly follow the JSON spec?

File: google-datacatalog-saphana-connector/src/google/datacatalog_connectors/saphana/datacatalog_cli.py [hunk 6]
```diff
@@ -0,0 +1,98 @@
...
+        parser.add_argument(
+            '--datacatalog-location-id',
+            help='Location ID to be used for your Google Cloud Datacatalog',
+            required=True)
+        parser.add_argument('--datacatalog-entry-group-id',
+                            help='Entry group ID to be used for your Google '
+                            'Cloud Datacatalog')
+        parser.add_argument(
+            '--saphana-host',
+            help='Your saphana server host, this is required even'
```
Comment (ricardolsmendes) [inline google-datacatalog-saphana-connector/src/google/datacatalog_connectors/saphana/datacatalog_cli.py:70]:
```suggestion
            help='Your SAP Hana server host, this is required even'
```

File: google-datacatalog-saphana-connector/src/google/datacatalog_connectors/saphana/datacatalog_cli.py [hunk 7]
```diff
@@ -0,0 +1,98 @@
...
+            ' for the raw_metadata_csv,'
+            ' so we are able to map the created entries'
+            ' resource with the sap hana host',
```
Comment (ricardolsmendes) [inline google-datacatalog-saphana-connector/src/google/datacatalog_connectors/saphana/datacatalog_cli.py:73]:
```suggestion
            ' resource with the SAP Hana host',
```

File: google-datacatalog-saphana-connector/src/google/datacatalog_connectors/saphana/datacatalog_cli.py [hunk 8]
```diff
@@ -0,0 +1,98 @@
...
+            required=True)
+        parser.add_argument('--saphana-user',
+                            help='Your saphana credentials user')
+        parser.add_argument('--saphana-pass',
+                            help='Your saphana credentials password')
```
Comment (ricardolsmendes) [inline google-datacatalog-saphana-connector/src/google/datacatalog_connectors/saphana/datacatalog_cli.py:78]:
```suggestion
                            help='Your SAP Hana credentials password')
```

File: google-datacatalog-saphana-connector/src/google/datacatalog_connectors/saphana/datacatalog_cli.py [hunk 9]
```diff
@@ -0,0 +1,98 @@
...
+        parser.add_argument('--saphana-database',
+                            help='Your saphana database name')
```
Comment (ricardolsmendes) [inline google-datacatalog-saphana-connector/src/google/datacatalog_connectors/saphana/datacatalog_cli.py:80]:
```suggestion
                            help='Your SAP Hana database name')
```

File: google-datacatalog-saphana-connector/tools/cleanup_datacatalog.py [hunk 10]
```diff
@@ -0,0 +1,122 @@
...
+                tag_template=tag_template_id),
+            force=True)
+        logging.info('--> Tag Template deleted: %s', tag_template_id)
+    except exceptions.GoogleAPICallError as e:
+        logging.warning('Exception deleting Tag Template: %s', str(e))
+
+
+def __parse_args():
+    parser = argparse.ArgumentParser(description='Command line to clean up all'
+                                     ' PostgreSQL metadata on Datacatalog')
```
Comment (ricardolsmendes) [inline google-datacatalog-saphana-connector/tools/cleanup_datacatalog.py:95]:
```suggestion
                                     ' SAP Hana metadata on Data Catalog')
```

File: google-datacatalog-saphana-connector/tools/cleanup_datacatalog.py [hunk 11]
```diff
@@ -0,0 +1,122 @@
...
+
+    parser.add_argument(
+        '--datacatalog-project-ids',
+        help='List of Google Cloud project IDs split by comma, '
+        'At least one must be specified',
+        required=True)
+    parser.add_argument(
+        '--datacatalog-location-id',
+        help='Location id which is the Region that your Datacatalog resides',
```
Comment (ricardolsmendes) [inline google-datacatalog-saphana-connector/tools/cleanup_datacatalog.py:104]:
```suggestion
        help='Location id which is the Region where your Data Catalog resides',
```

File: google-datacatalog-saphana-connector/src/google/datacatalog_connectors/saphana/scrape/metadata_enricher.py [hunk 12]
```diff
@@ -0,0 +1,65 @@
...
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+from google.datacatalog_connectors.rdbms.scrape import \
+    config_constants, metadata_enricher
+
+import re
```
Comment (ricardolsmendes) [inline google-datacatalog-saphana-connector/src/google/datacatalog_connectors/saphana/scrape/metadata_enricher.py:20]: Please import the built-in libraries first to keep it consistent with other files 😃.

**Review by ricardolsmendes — CHANGES_REQUESTED** [review]
@mesmacosta , the code LGTM. I've left some comments to improve the docs and follow standards. Back to you! 🏓

File: google-datacatalog-saphana-connector/setup.py (hunk 1 above)
Comment (mesmacosta) [reply google-datacatalog-saphana-connector/setup.py:45]: Cloud shell uses 3.7 by default, so I'd suggest changing the README file instead.

File: google-datacatalog-saphana-connector/README.md [hunk 13]
```diff
@@ -0,0 +1,269 @@
...
+
+## 1. Installation
+
+Install this library in a [virtualenv][1] using pip. [virtualenv][1] is a tool to
+create isolated Python environments. The basic problem it addresses is one of
+dependencies and versions, and indirectly permissions.
+
+With [virtualenv][1], it's possible to install this library without needing system
+install permissions, and without clashing with the installed system
+dependencies. Make sure you use Python 3.6+. This connector is tested in versions 3.6, 3.7, 3.8.
```
Comment (ricardolsmendes) [inline google-datacatalog-saphana-connector/README.md:57]: "Please stick to 80-character line wraps as much as you can." 😃

File: google-datacatalog-saphana-connector/src/google/datacatalog_connectors/saphana/datacatalog_cli.py [hunk 14]
```diff
@@ -0,0 +1,98 @@
...
+                            'config/metadata_definition.json')
+
+    def _get_query_path(self, args):
+        return os.path.join(os.path.dirname(os.path.abspath(__file__)),
+                            'config/metadata_query.sql')
+
+    def _parse_args(self, argv):
+        parser = argparse.ArgumentParser(
+            description='Command line to sync SAP Hana '
+            'metadata to Datacatalog')
```
Comment (ricardolsmendes) [inline google-datacatalog-saphana-connector/src/google/datacatalog_connectors/saphana/datacatalog_cli.py:56]:
```suggestion
            'metadata into Google Data Catalog')
```

File: google-datacatalog-saphana-connector/src/google/datacatalog_connectors/saphana/datacatalog_cli.py [hunk 15]
```diff
@@ -0,0 +1,98 @@
...
+
+        parser.add_argument('--datacatalog-project-id',
+                            help='Your Google Cloud project ID',
+                            required=True)
+        parser.add_argument(
+            '--datacatalog-location-id',
+            help='Location ID to be used for your Google Cloud Datacatalog',
```
Comment (ricardolsmendes) [inline google-datacatalog-saphana-connector/src/google/datacatalog_connectors/saphana/datacatalog_cli.py:63]:
```suggestion
            help='Location ID to be used for your Google Data Catalog',
```

File: google-datacatalog-saphana-connector/src/google/datacatalog_connectors/saphana/datacatalog_cli.py [hunk 16]
```diff
@@ -0,0 +1,98 @@
...
+            required=True)
+        parser.add_argument('--datacatalog-entry-group-id',
+                            help='Entry group ID to be used for your Google '
+                            'Cloud Datacatalog')
```
Comment (ricardolsmendes) [inline google-datacatalog-saphana-connector/src/google/datacatalog_connectors/saphana/datacatalog_cli.py:67]:
```suggestion
                            'Data Catalog')
```

**Review by ricardolsmendes — CHANGES_REQUESTED** [review]
A few more nitpicks... 🏓

**Review by ricardolsmendes — APPROVED** [review]
LGTM. Thanks!

**Pull request merged**